core/src/main/java/org/elasticsearch/bootstrap/FilePermissionUtils.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.bootstrap;
+
+import org.elasticsearch.common.SuppressForbidden;
+
+import java.io.FilePermission;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.security.Permissions;
+
+public class FilePermissionUtils {
+
+    /** no instantiation */
+    private FilePermissionUtils() {}
+
+    private static final boolean VERSION_IS_AT_LEAST_JAVA_9 = JavaVersion.current().compareTo(JavaVersion.parse("9")) >= 0;
+
+    /**
+     * Add access to single file path
+     * @param policy current policy to add permissions to
+     * @param path the path itself
+     * @param permissions set of file permissions to grant to the path
+     */
+    @SuppressForbidden(reason = "only place where creating Java-9 compatible FilePermission objects is possible")
+    public static void addSingleFilePath(Permissions policy, Path path, String permissions) throws IOException {
+        policy.add(new FilePermission(path.toString(), permissions));
+        if (VERSION_IS_AT_LEAST_JAVA_9 && Files.exists(path)) {
+            // Java 9 FilePermission model requires this due to the removal of pathname canonicalization,
+            // see also https://github.com/elastic/elasticsearch/issues/21534
+            Path realPath = path.toRealPath();
+            if (path.toString().equals(realPath.toString()) == false) {
+                policy.add(new FilePermission(realPath.toString() + realPath.getFileSystem().getSeparator() + "-", permissions));
```

Comment:
Shouldn't the dash only be for a directory? But this method is for adding a single file?